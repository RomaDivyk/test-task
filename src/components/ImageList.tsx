import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { MediaCart } from './index';

import { StyledImageList } from '../styled/StyledImageList';

import { getDataFetch } from '../redux/ducks/imagesDuck';

export interface Data {
  id: string;
  img: string;
  comments: string[];
}
export interface FetchData {
  data: Data[];
}

export type StateType = {
  fetchData: FetchData;
  isLoading: boolean;
};

const ImageList = () => {
  const data = useSelector((state: StateType) => state.fetchData.data);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getDataFetch());
  }, [dispatch]);

  return (
    <StyledImageList data-testid="imageListId">
      <MediaCart data={data} />
    </StyledImageList>
  );
};

export default ImageList;
