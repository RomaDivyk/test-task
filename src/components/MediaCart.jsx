import React, { useState } from 'react';

import { Container, Grid, CardMedia } from '@mui/material';

import { ModalWindow } from './index';

import { StyledMediaCart } from '../styled/StyledMediaCart';

const MediaCard = ({ data }) => {
  const [open, setOpen] = useState(false);
  const [id, setId] = useState(null);

  const handleOpen = (event) => {
    setId(event.target.id);
    setOpen(true);
  };
  const handleClose = () => setOpen(false);

  const cards = data.map((element) => (
    <Grid item key={element.id} xs={12} sm={6} md={4}>
      <StyledMediaCart>
        <CardMedia
          onClick={handleOpen}
          id={element.id}
          component="img"
          image={element.img}
          alt={`img doesn't download`}
        ></CardMedia>
      </StyledMediaCart>
    </Grid>
  ));

  return (
    <>
      <Container>
        <Grid container spacing={4}>
          {cards}
        </Grid>
      </Container>

      <ModalWindow id={id} data={data} open={open} onClose={handleClose} />
    </>
  );
};

export default MediaCard;
