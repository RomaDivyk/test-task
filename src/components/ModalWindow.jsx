import { useState } from 'react';
import { Modal, Box, CardMedia, TextField, Button, Container, Grid } from '@mui/material';

import { StyledModalWindow, Line, styleModal } from '../styled/StyledModalWindow';

import { sendData } from '../api';

const ModalWindow = ({ id, data, open, onClose }) => {
  const [modalObj, setModalObj] = useState({
    id: null,
    img: '',
    comments: '',
  });
  const [inputData, setInputData] = useState({
    name: '',
    comment: '',
  });

  if (id !== null) {
    const modalResult = data.filter((element) => element.id === id);

    modalObj.id = modalResult[0].id;
    modalObj.img = modalResult[0].img;
    modalObj.comments = modalResult[0].comments.map((comment, index) => <Line key={index}>{comment}</Line>);
  }

  return (
    <StyledModalWindow>
      <Modal open={open} onClose={onClose}>
        <Container sx={styleModal}>
          <Grid container spacing={2}>
            <Grid item xs={10} sm={6.5} md={6} lg={4.5} xl={4} sx={{ maxHeight: '100%' }}>
              <CardMedia image={modalObj.img} alt="empty" component="img" sx={{ maxHeight: '80%', marginBottom: 3 }} />
              <Box>{modalObj.comments}</Box>
            </Grid>

            <Grid item xs={10} sm={6} md={4}>
              <Box
                sx={{
                  display: 'flex',
                  flexDirection: 'column',
                  width: '100%',
                  height: 'auto',
                }}
              >
                <TextField
                  onChange={(e) => setInputData({ ...inputData, name: e.target.value })}
                  id="filled-basic-name"
                  label="Name"
                  variant="filled"
                  margin="normal"
                />
                <TextField
                  onChange={(e) => setInputData({ ...inputData, comment: e.target.value })}
                  id="filled-basic-comment"
                  label="Comment"
                  variant="filled"
                  margin="normal"
                />
                <Button variant="contained" onClick={() => sendData(inputData)}>
                  Leave a comment{' '}
                </Button>
              </Box>
            </Grid>
          </Grid>
        </Container>
      </Modal>
    </StyledModalWindow>
  );
};

export default ModalWindow;
