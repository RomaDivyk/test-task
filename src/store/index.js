import { configureStore } from "@reduxjs/toolkit";
import consoleSlice from "./console_slice";

const store = configureStore({
  reducer: {
    cons: consoleSlice,
  },
});

export default store;
