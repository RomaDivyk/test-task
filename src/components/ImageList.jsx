/* eslint-disable no-unused-vars */
import { useState, useEffect } from "react";

import { fetchData } from "../api";

import MediaCart from "./MediaCart";

const ImageList = () => {
  const [data, setData] = useState([]);
  const [error, setError] = useState("");

  const url = `https://demo1353770.mockable.io/images`;

  useEffect(() => {
    fetchData(url).then((res) => {
      setData(res);
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <>
      {<MediaCart data={data} />}
      {error && <h1>{error}</h1>}
    </>
  );
};

export default ImageList;
