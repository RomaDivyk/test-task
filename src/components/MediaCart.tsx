import { useState } from 'react';

import { Container, Grid, CardMedia } from '@mui/material';

import { ModalWindow } from './index';
import { FetchData } from './ImageList';

import { StyledMediaCart } from '../styled/StyledMediaCart';

const MediaCard = ({ data }: FetchData) => {
  const [open, setOpen] = useState(false);
  const [id, setId] = useState<string>('');

  const handleOpen: React.MouseEventHandler<HTMLImageElement> = (event) => {
    setId((event.target as Element).id);
    setOpen(true);
  };
  const handleClose = (): void => setOpen(false);

  const cards = data.map((element) => (
    <Grid item key={element.id} xs={12} sm={6} md={4}>
      <StyledMediaCart>
        <CardMedia
          onClick={handleOpen}
          id={element.id}
          component="img"
          image={element.img}
          alt={`img doesn't download`}
        ></CardMedia>
      </StyledMediaCart>
    </Grid>
  ));

  return (
    <>
      <Container data-testid="mediaCartId">
        <Grid container spacing={4}>
          {cards}
        </Grid>
      </Container>

      <ModalWindow id={id} data={data} open={open} onClose={handleClose} />
    </>
  );
};

export default MediaCard;
