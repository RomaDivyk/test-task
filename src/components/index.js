/// check

import ImageList from "./ImageList";
import MediaCart from "./MediaCart";
import ModalWindow from "./ModalWindow";

export { ImageList, MediaCart, ModalWindow };
