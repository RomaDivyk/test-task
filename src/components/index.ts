import ImageList from './ImageList';
import MediaCart from './MediaCart';
import ModalWindow from './ModalWindow';
import Form from './auth/Form';
import ReactGridTable from './ReactGridTable';
import Charts from './Charts';

export { ImageList, MediaCart, ModalWindow, Form, ReactGridTable, Charts };
