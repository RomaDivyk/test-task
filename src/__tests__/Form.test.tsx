import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import store from '../redux';

import { Form } from '../components';

describe('testing Form component', () => {
  it('should match a snapshot in Form component', () => {
    const wrapper = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Form />
        </BrowserRouter>
      </Provider>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  const mainRender = () => {
    render(
      <Provider store={store}>
        <Form />
      </Provider>,
      { wrapper: BrowserRouter }
    );
  };

  it('should contain heading element', () => {
    mainRender();
    const headingElement = screen.getByRole('heading', { name: 'REGISTRATION' });
    expect(headingElement).toBeInTheDocument();
  });
  it('should render First Name input', () => {
    mainRender();
    const firstNameInputElement = screen.getByRole('textbox', { name: /first name/i });
    expect(firstNameInputElement).toBeInTheDocument();
  });
  it('should render Last Name input', () => {
    mainRender();
    const lastNameInputElement = screen.getByRole('textbox', { name: /last name/i });
    expect(lastNameInputElement).toBeInTheDocument();
  });
  it('should render Email Addres input', () => {
    mainRender();
    const emailAddresInputElement = screen.getByRole('textbox', { name: /email addres/i });
    expect(emailAddresInputElement).toBeInTheDocument();
  });
  it('should render Password input', () => {
    mainRender();
    const passwordInputElement = screen.getByLabelText('Password');
    expect(passwordInputElement).toBeInTheDocument();
  });
  it('should render Repeat password input', () => {
    mainRender();
    const repeatPasswordInputElement = screen.getByLabelText('Repeat password');
    expect(repeatPasswordInputElement).toBeInTheDocument();
  });
  it('should render button and click on it', () => {
    mainRender();
    const buttonElement = screen.getByRole('button', { name: /submit/i });
    userEvent.click(buttonElement);
    expect(buttonElement).toBeDefined();
  });
});
