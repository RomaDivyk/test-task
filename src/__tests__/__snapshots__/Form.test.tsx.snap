// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Form component should match a snapshot in Form component 1`] = `
Array [
  <h1>
    AUTH PAGE
  </h1>,
  <h2>
    REGISTRATION
  </h2>,
  <div
    className="sc-eDLKkx ereGeJ"
  >
    <form
      className="sc-jTQCzO eeqcqO"
      onSubmit={[Function]}
    >
      <div
        className="MuiFormControl-root MuiFormControl-marginNormal MuiTextField-root css-1rv3ei8-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-filled css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="firstName"
          id="firstName-label"
        >
          First Name
        </label>
        <div
          className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-batk84-MuiInputBase-root-MuiFilledInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiFilledInput-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
            disabled={false}
            id="firstName"
            name="firstName"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiFormControl-marginNormal MuiTextField-root css-1rv3ei8-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-filled css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="lastName"
          id="lastName-label"
        >
          Last Name
        </label>
        <div
          className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-batk84-MuiInputBase-root-MuiFilledInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiFilledInput-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
            disabled={false}
            id="lastName"
            name="lastName"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiFormControl-marginNormal MuiTextField-root css-1rv3ei8-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-filled css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="email"
          id="email-label"
        >
          Email Address
        </label>
        <div
          className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-batk84-MuiInputBase-root-MuiFilledInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiFilledInput-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
            disabled={false}
            id="email"
            name="email"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="email"
            value=""
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiFormControl-marginNormal MuiTextField-root css-1rv3ei8-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-filled css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="password"
          id="password-label"
        >
          Password
        </label>
        <div
          className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-batk84-MuiInputBase-root-MuiFilledInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiFilledInput-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
            disabled={false}
            id="password"
            name="password"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiFormControl-marginNormal MuiTextField-root css-1rv3ei8-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-filled css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="repeatPassword"
          id="repeatPassword-label"
        >
          Repeat password
        </label>
        <div
          className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-batk84-MuiInputBase-root-MuiFilledInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiFilledInput-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
            disabled={false}
            id="repeatPassword"
            name="repeatPassword"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="password"
            value=""
          />
        </div>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>,
]
`;
